storage-proofs/core/src/gadgets/insertion.rs

```diff
@@ -69,6 +86,172 @@ pub fn insert<E: Engine, CS: ConstraintSystem<E>>(
     Ok(result)
 }
 
+pub fn insert_4<E: Engine, CS: ConstraintSystem<E>>(
```

Comment:
when do we use quads?

Reply:
Quads are needed any time we need to build a tree which cannot be constructed either with:
- only oct trees
- only oct and binary trees

Put differently, we need them whenever log N mod 3 = 2, where N is the number of 32-byte leaves. Although not currently supported, this would include either 16 or 128GiB sectors. Since we have only currently generated parameters for 32 and 64GiB sectors, quad hashing is dormant. But it's important that we preserve the freedom to support any power of two sector size — so the audit needs to cover quad hashes.